crypto/s2n_certificate.c

```diff
@@ -532,3 +532,56 @@ s2n_cert_private_key *s2n_cert_chain_and_key_get_private_key(struct s2n_cert_cha
     PTR_ENSURE_REF(chain_and_key);
     return chain_and_key->private_key;
 }
+
+int s2n_get_cert_chain_length(struct s2n_cert_chain_and_key *chain_and_key, uint32_t *cert_length)
+{
+    POSIX_ENSURE_REF(chain_and_key);
+
+    struct s2n_cert *head_cert = chain_and_key->cert_chain->head;
+    POSIX_ENSURE_REF(head_cert);
+    *cert_length = 1;
+    struct s2n_cert *next_cert = head_cert->next;
+    while (next_cert != NULL) {
+        *cert_length += 1;
+        next_cert = next_cert->next;
+    }
+
+    return S2N_SUCCESS;
+}
+
+int s2n_get_cert_from_cert_chain(struct s2n_cert_chain_and_key *chain_and_key, struct s2n_cert **out_cert,
+                                 uint32_t cert_idx)
+{
+    POSIX_ENSURE_REF(chain_and_key);
+    POSIX_ENSURE_REF(out_cert);
+
+    struct s2n_cert *cur_cert = chain_and_key->cert_chain->head;
+    POSIX_ENSURE_REF(cur_cert);
+    uint32_t counter = 0;
+
+    struct s2n_cert *next_cert = cur_cert->next;
+
+    while ((next_cert != NULL) && (counter < cert_idx)) {
+        cur_cert  = next_cert;
+        next_cert = next_cert->next;
+        counter++;
+    }
+
+    POSIX_ENSURE_EQ(counter, cert_idx);
+    POSIX_ENSURE_REF(cur_cert);
+    *out_cert = cur_cert;
+
+    return S2N_SUCCESS;
+}
+
+int s2n_get_cert_der(struct s2n_cert *cert, uint8_t *out_cert_der, uint32_t *cert_length)
+{
+    POSIX_ENSURE_REF(cert);
+    POSIX_ENSURE_REF(out_cert_der);
+    POSIX_ENSURE_REF(cert_length);
+
+    *cert_length = cert->raw.size;
+    POSIX_CHECKED_MEMCPY(out_cert_der, cert->raw.data, cert->raw.size);
```

Comment:
Should we also require the size of the in-buffer be supplied in `cert_length` and then ensure that it is greater than or equal to `cert->raw.size`?

Reply:
We could do that, but I dont see any real benefit out of it as `cert->raw` is the`s2n_blob` object that holds the `in-buffer` and its size.

Reply:
This question is under the assumption that we copy into user-controlled memory. In this case it is useful for them to tell us how much memory they gave us so that we know we aren't copying too much data into it.